internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -170,7 +186,40 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 	}
 
 	det.SetState(details.StateExtracting)
-	// TODO: add check that no archive files end up in the current versioned home
+
+	metadata, err := u.getPackageMetadata(archivePath)
+	if err != nil {
+		return nil, fmt.Errorf("reading metadata for elastic agent version %s package %q: %w", version, archivePath, err)
+	}
+
+	currentVersion := agentVersion{
+		version:  release.Version(),
+		snapshot: release.Snapshot(),
+		hash:     release.Commit(),
+	}
+
+	newVersion := agentVersion{}
+	if metadata.manifest != nil {
+		packageDesc := metadata.manifest.Package
+		newVersion.version = packageDesc.Version
+		newVersion.snapshot = packageDesc.Snapshot
+	} else {
+		// extract version info from the version string (we can ignore parsing errors as it would have never passed the download step)
+		parsedVersion, _ := agtversion.ParseVersion(version)
+		newVersion.version = strings.TrimSuffix(parsedVersion.VersionWithPrerelease(), "-SNAPSHOT")
+		newVersion.snapshot = parsedVersion.IsSnapshot()
+	}
+	newVersion.hash = metadata.hash
+
+	u.log.Debugw("Comparing current and new agent version", "current_version", currentVersion, "new_version", newVersion)
+
+	if currentVersion == newVersion {
```

Comment:
This looks like it works by inspection, but it feels like this section could be unit tested. Unless I am missing that this is already covered somewhere else, or the intent is to rely only on the integration tests.

Cases I think of that could be tested:
1. Ensuring that we allow upgrade when only the hash differs. This should be covered by our integration test that upgrades to the same snapshot version.
2. Ensuring we cover the case where one version is from the manifest and the other is from a legacy agent. Again I think the upgrade from 7.17 unit test covers this.

Reply:
It can surely be unit tested, I didn't do it because I considered the comparison of 2 structs "too simple to fail".
If we a mock package  and a mock agent install into play then it could become interesting but it is also true that it should be pretty simple to have an integration test for this. 
I do not have a strong preference between the 2 options

Reply:
I agree with @cmacknz on adding a unit test. I don't think he meant specifically to test this one line, I think the whole function in general should have some coverage. Testing the manifest usage and string parsing before this line to ensure that those are filling in the data as expected and the expected error is returned or not returned.

Reply:
added in c9b41aae094713fd1ad7a13615644be63fe0840b